File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    
    answer = -1
    def dfs(c, k, dun):
        nonlocal answer
        
        for i, d in enumerate(dun):
            if k >= d[0]:
                nd = dun.copy()
                nd.pop(i)
                mx = dfs(c + 1, k - d[1], nd)
            
                answer = max(answer, mx)
        return c
    
    dfs(0 ,k, dungeons)
    
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    int answer = 0;
    boolean[] visited;

    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        dfs(0, k, dungeons);
        return answer;
    }

    public void dfs(int count, int k, int[][] dungeons) {
        answer = Math.max(answer, count);

        for (int i = 0; i < dungeons.length; i++) {
            if (!visited[i] && k >= dungeons[i][0]) {
                visited[i] = true;
                dfs(count + 1, k - dungeons[i][1], dungeons);
                visited[i] = false;
            }
        }
    }
}
